extensions/libevent.c

```diff
@@ -0,0 +1,318 @@
+#define FUNC_NAME "primitive-event-wake"
+{
+  int c_fd;
+  char zero = 0;
+
+  c_fd = scm_to_int (wakefd);
+
+  if (write (c_fd, &zero, 1) <= 0 && errno != EWOULDBLOCK && errno != EAGAIN)
+    SCM_SYSERROR;
+
+  return SCM_UNSPECIFIED;
+}
+#undef FUNC_NAME
+
+static SCM
+scm_primitive_create_event_base (SCM eventsv)
+#define FUNC_NAME "primitive-create-event-base"
+{
+  struct libevt_data *libevt;
+  struct wait_data *data;
+  struct event_base *base;
+
+  if ((base = event_base_new ()) == NULL)
+    scm_syserror ("couldn't instantiate event_base");
+
+  libevt = (struct libevt_data *) scm_gc_malloc (sizeof (struct libevt_data),
+                                                 "libevt_data");
+  libevt->base = base;
+
+  data = (struct wait_data *) scm_gc_malloc (sizeof (struct wait_data),
+                                             "wait_data");
+  data->rv = 0;
+  data->events = (struct event_data *) SCM_BYTEVECTOR_CONTENTS (eventsv);
+  data->maxevents = SCM_BYTEVECTOR_LENGTH (eventsv) / sizeof (struct event_data);
+
+  return scm_list_2 (scm_from_pointer (libevt, free_libevt),
+                     scm_from_pointer (data, free_wait_data));
+}
+#undef FUNC_NAME
+
+static SCM
+scm_primitive_add_event (SCM lst, SCM fd, SCM ev)
+#define FUNC_NAME "primitive-add-event"
+{
+  int c_fd;
+  short c_ev;
+  struct event *event;
+  struct wait_data *data;
+  struct libevt_data *libevt;
+
+  c_fd = scm_to_int (fd);
+  c_ev = scm_to_short (ev);
+
+  libevt =
+    (struct libevt_data *) scm_to_pointer (scm_list_ref (lst, scm_from_int (0)));
+  data =
+    (struct wait_data *) scm_to_pointer (scm_list_ref (lst, scm_from_int (1)));
+
+  event = event_new (libevt->base, c_fd, c_ev, cb_func, data);
+  event_add (event, NULL);
+
+  return scm_from_pointer (event, NULL);
+}
+#undef FUNC_NAME
+
+static SCM
+scm_primitive_remove_event (SCM lst, SCM scm_event)
+#define FUNC_NAME "primitive-remove-event"
+{
+  struct event *event = scm_to_pointer(scm_event);
+
+  event_del (event);
+
+  return SCM_UNSPECIFIED;
+}
+#undef FUNC_NAME
+
+static SCM
+scm_primitive_resize (SCM lst, SCM eventsv)
+#define FUNC_NAME "primitive-resize"
+{
+  struct wait_data *data;
+
+  data =
+    (struct wait_data *) scm_to_pointer (scm_list_ref (lst, scm_from_int (1)));
+
+  data->events = (struct event_data *) SCM_BYTEVECTOR_CONTENTS (eventsv);
+  data->maxevents = SCM_BYTEVECTOR_LENGTH (eventsv) / sizeof (struct event_data);
+
+  return SCM_UNSPECIFIED;
+}
+#undef FUNC_NAME
+
+static uint64_t time_units_per_microsec;
+
+static void*
+run_event_loop (void *p)
+{
+  int microsec = 0;
+  struct timeval tv;
+
+  struct loop_data *data = p;
+
+  if (data->timeout < 0)
+    microsec = -1;
+  else if (data->timeout >= 0)
+    {
+      microsec = data->timeout / time_units_per_microsec;
+      tv.tv_sec = 0;
+      tv.tv_usec = microsec;
+    }
+
+  if (microsec >= 0)
+    event_base_loopexit (data->base, &tv);
+
+  event_base_loop (data->base, EVLOOP_ONCE);
```

Comment:
Likewise, according to the documentation, this can return an error value.

Reply:
done.